internal/datacoord/channel_manager.go

```diff
@@ -868,30 +597,171 @@ func (c *ChannelManager) getChannelByNodeAndName(nodeID UniqueID, channelName st
 	return ret
 }
 
-func (c *ChannelManager) getNodeIDByChannelName(chName string) (bool, UniqueID) {
+// checkOldNodes processes the existing watch channels when starting up.
+// ToWatch         re-notify and start check loop
+// WatchSuccess    ignore
+// WatchFail       set to ToRelease
+// ToRelase        re-notify and start check loop
+// ReleaseSuccess  Reassign
+// ReleaseFail     CleanupAndReassign
+func (c *ChannelManager) checkOldNodes(nodes []UniqueID) error {
+	// Load all the watch infos before processing
+	nodeWatchInfos := make(map[UniqueID][]*datapb.ChannelWatchInfo)
+	for _, nodeID := range nodes {
+		watchInfos, err := c.store.GetNodeWatchInfos(nodeID)
+		if err != nil {
+			return err
+		}
+		nodeWatchInfos[nodeID] = watchInfos
+	}
+
+	for nodeID, watchInfos := range nodeWatchInfos {
+		for _, info := range watchInfos {
+			channelName := info.GetVchan().GetChannelName()
+
+			log := log.With(
+				zap.Int64("nodeID", nodeID),
+				zap.String("watch state", info.GetState().String()),
+				zap.String("channel", channelName),
+			)
+			ch := c.getChannelByNodeAndName(nodeID, channelName)
+			if ch == nil {
+				log.Warn("fail to find matching nodeID with channel")
+				continue
+			}
+
+			log.Info("processing watch info")
+
+			switch info.GetState() {
+			case datapb.ChannelWatchState_ToWatch, datapb.ChannelWatchState_Uncomplete:
+				c.notifyAndCheck(&ChannelOp{
+					Type:              Add,
+					NodeID:            nodeID,
+					Channels:          []*channel{ch},
+					ChannelWatchInfos: []*datapb.ChannelWatchInfo{info},
+				})
+
+			case datapb.ChannelWatchState_WatchFailure:
+				if err := c.Release(nodeID, channelName); err != nil {
+					log.Warn("fail to release", zap.Error(err))
+					return err
+				}
+
+			case datapb.ChannelWatchState_ToRelease:
+				c.notifyAndCheck(&ChannelOp{
+					Type:              Add,
```

Comment:
Should it be Delete?

Reply:
No, delete will actually delete this kv, Delete will only be used in transactions of Delete + Add or when fully removing a channel. Not when try to balance.